### DISCUSSION TITLE
memory prefault

### DISCUSSION THREAD

**Discussion by Maggie262** [cloud-hypervisor/cloud-hypervisor#2874] [Q&A]
Hello, I reading cloud-hypervisor's `vmm/src/memory_manager.rs` module, and notice that `MAP_POPULATE` is used to prefault pages. I have questions about that: 

1. this `MAP_POPULATE` flag is not only valid for private mappings, but also for file mappings and share mappings, is that right?
2. With `MAP_POPULATE` flag, host kernel will prefault the pages mapped, but mmap syscall will not fail if the mapping cannot be populated ([linux man page](https://man7.org/linux/man-pages/man2/mmap.2.html)). Then in cloud-hypervisor, how can we know if prefault is successful or not?

**Comment by sboeuf** [comment]
Hi @Maggie262 

>this MAP_POPULATE flag is not only valid for private mappings, but also for file mappings and share mappings, is that right?

`MAP_POPULATE` is only valid for private mappings. But a private mapping can be either based on anonymous pages or a real file.

>With MAP_POPULATE flag, host kernel will prefault the pages mapped, but mmap syscall will not fail if the mapping cannot be populated (linux man page). Then in cloud-hypervisor, how can we know if prefault is successful or not?

We don't have any way of knowing about this. But we'll see an error from the guest when it tries to access this memory since the pagefault might not bring the page in.

**Reply by Maggie262** [reply]
Thanks for your kind help！ @sboeuf 

As for the first question, you mentioned that `MAP_POPULATE` is only valid for private mapping. But When I call `mmap` syscall and set `flags = MAP_ANONYMOUS | MAP_POPULATE`, `prot= PROT_READ | PROT_WRITE`, the perf data shows that it also prefaults all pages. The **RES** of this process is also equal to the mapping size. 

I still wondor that if `MAP_POPULATE` is valid for shared mappings. This question is also mentioned [there](https://stackoverflow.com/questions/23502361/linux-mmap-with-map-populate-man-page-seems-to-give-wrong-info) .

**Reply by sboeuf** [reply]
Have you tried `flags = MAP_POPULATE | MAP_SHARED`? This should fail according to the man page.

**Reply by Maggie262** [reply]
My error, I have set `MAP_SHARED` flag in the previous test. The result shows that it also prefaults all pages.

**Reply by sboeuf** [reply]
Well then the man page is not up to date :)

**Reply by Maggie262** [reply]
@sboeuf I also notice that qemu does not use `MAP_POPULATE` flag to prefault pages, but writes every page with multi-threads. Why qemu choose to use this method?

**Reply by sboeuf** [reply]
I don't know, I guess that is their way of checking that they could populate correctly the memory.

**Reply by Maggie262** [reply]
Thanks for your help  : )
